includes/class-wc-regenerate-images.php

```diff
@@ -0,0 +1,195 @@
+
+		if ( ! file_exists( $original_image_file_path ) || ! getimagesize( $original_image_file_path ) ) {
+			return $image;
+		}
+
+		$info = pathinfo( $original_image_file_path );
+		$ext = $info['extension'];
+		list( $orig_w, $orig_h ) = getimagesize( $original_image_file_path );
+		// Get image size after cropping.
+		$image_size = wc_get_image_size( $size );
+		$dimensions = image_resize_dimensions( $orig_w, $orig_h, $image_size['width'], $image_size['height'], $image_size['crop'] );
+		if ( ! $dimensions || ! is_array( $dimensions ) ) {
+			return $image;
+		}
+
+		$dst_w        = $dimensions[4];
+		$dst_h        = $dimensions[5];
+		$suffix       = "{$dst_w}x{$dst_h}";
+		$dst_rel_path = str_replace( '.' . $ext, '', $original_image_file_path );
+		$destfilename = "{$dst_rel_path}-{$suffix}.{$ext}";
+
+		// If the file is already there perhaps just load it.
+		if ( file_exists( $destfilename ) ) {
+			return array(
+				0 => str_replace( $wp_uploads_dir, $wp_uploads_url, $destfilename ),
+				1 => $image_size['width'],
+				2 => $image_size['height'],
+			);
+		}
+
+		// Lets resize the image if it does not exist yet.
+		$editor = wp_get_image_editor( $original_image_file_path );
+		if ( is_wp_error( $editor ) || is_wp_error( $editor->resize( $image_size['width'], $image_size['height'], $image_size['crop'] ) ) ) {
+			return $image;
+		}
+		$resized_file = $editor->save();
+		if ( ! is_wp_error( $resized_file ) ) {
+			$img_url = str_replace( $wp_uploads_dir, $wp_uploads_url, $resized_file['path'] );
+			return array(
+				0 => $img_url,
+				1 => $image_size['width'],
+				2 => $image_size['height'],
+			);
+		}
+
+		// Lets just add this here as a fallback.
+		return $image;
+	}
+
+	/**
+	 * Check if we should regenerate the product images using a job queue in the background.
+	 *
+	 * @param string $option Option name.
+	 * @param mixed  $old_value Old option value.
+	 * @param mixed  $new_value New option value.
+	 */
+	public static function maybe_regenerate_images_background( $option, $old_value, $new_value ) {
+		global $wpdb;
+
+		if ( ! apply_filters( 'woocommerce_background_image_regeneration', true ) ) {
+			return;
+		}
+		if ( ! in_array( $option, array( 'woocommerce_thumbnail_cropping', 'woocommerce_thumbnail_image_width', 'woocommerce_single_image_width' ), true ) ) {
+			return;
+		}
+
+		if ( $new_value === $old_value ) {
+			return;
+		}
+
+		// If we made it till here, it means that image settings have changed.
+		// First lets cancel existing running queue to avoid running it more than once.
+		self::$background_process->cancel_process();
+
+		// Now lets find all product image attachments IDs and pop them onto the queue.
+		$images = $wpdb->get_results( // @codingStandardsIgnoreLine
+			"SELECT ID
+			FROM $wpdb->posts
+			WHERE post_type = 'attachment'
+			AND post_mime_type LIKE 'image/%'
+			ORDER BY ID DESC"
+		);
+		foreach ( $images as $image ) {
+			self::$background_process->push_to_queue( array(
```

Comment:
What performance implications does a large queue have when there are lots of images?

Reply:
Because it is a queue it will execute in order, so it will not flood the site with concurrent requests. It will just run each resize one after the other. It might run a bit longer but it will not overload the site with requests.